overlord/ifacestate/hotplug.go

```diff
@@ -30,3 +35,98 @@ func (m *InterfaceManager) HotplugDeviceAdded(devinfo *hotplug.HotplugDeviceInfo
 // HotplugDeviceRemoved gets called when a device is removed from the system.
 func (m *InterfaceManager) HotplugDeviceRemoved(devinfo *hotplug.HotplugDeviceInfo) {
 }
+
+// ensureUniqueName modifies proposedName so that it's unique according to isUnique predicate.
+// Uniqueness is achieved by appending a numeric suffix, or increasing existing suffix.
+func ensureUniqueName(proposedName string, isUnique func(string) bool) string {
+	// if the name is unique right away, do nothing
+	if isUnique(proposedName) {
+		return proposedName
+	}
+
+	suffixNumValue := 0
+	prefix := strings.TrimRightFunc(proposedName, unicode.IsDigit)
+	if prefix != proposedName {
+		suffixNumValue, _ = strconv.Atoi(proposedName[len(prefix):])
+	}
+	prefix = strings.TrimRight(prefix, "-")
+
+	// increase suffix value until we have a unique name
+	for {
+		suffixNumValue++
+		proposedName = fmt.Sprintf("%s%d", prefix, suffixNumValue)
+		if isUnique(proposedName) {
+			return proposedName
+		}
+	}
+}
+
+const maxGenerateSlotNameLen = 20
+
+// makeSlotName sanitizes a string to make it a valid slot name that
+// passes validation rules implemented by ValidateSlotName (see snap/validate.go):
+// - only lowercase letter, digits and dashes are allowed
+// - must start with a letter
+// - no double dashes, cannot end with a dash.
+// In addition names are truncated not to exceed maxGenerateSlotNameLen characters.
+func makeSlotName(s string) string {
+	var out []rune
+	var charCount int
```

Comment:
How's charCount different from len(out)?

Reply:
Good point, the difference was subtle and not worth keeping I think - the dashes would not affect character count, so you could actually get a string longer than max length if there were dashes there. Simplified.